apps/readingchallenge/reading_challenge.star

```diff
@@ -0,0 +1,158 @@
+            render.Padding(child = render.Image(src = ONE_CLOSED_BOOK_ICON, height = 15, width = 12), pad = (1, 0, 0, 0)),
+        ])
+
+    elif progress == 12:
+        images = render.Row(main_align = "center", cross_align = "end", expanded = True, children = [
+            render.Image(src = FIVE_BOOK_ICON, height = 12, width = 18),
+            render.Image(src = THREE_BOOK_ICON, height = 22, width = 20),
+            render.Image(src = FIVE_BOOK_ICON, height = 16, width = 10),
+        ])
+    else:
+        images = render.Row(main_align = "center", cross_align = "end", expanded = True, children = [
+            render.Image(src = FIVE_BOOK_ICON, height = 15, width = 20),
+            render.Image(src = FIVE_BOOK_ICON, height = 14, width = 20),
+            render.Image(src = FIVE_BOOK_ICON, height = 12, width = 18),
+            render.Image(src = FIVE_BOOK_ICON, height = 16, width = 10),
+        ])
+
+    return images
+
+def main(config):
+    CHALLENGE_ID = secret.decrypt("AV6+xWcE/k5WZk3zYbxM/fuI6z4DJ5oQUNrHWbIu8HBOyeubTAzYcB9DxEMAWRoxDcryQFohadBlZneiSq94urEiiG53bpXexcibIvOyKOSH3Vi3qEYaY1vVXUxHQOmyk5odoCU9UtLX850GqOI=")
+
+    if CHALLENGE_ID == None:
+        CHALLENGE_ID = config.str("user_challenge_id", "0")
+
+    progress_cached = cache.get("progress")
+    goal_cached = cache.get("goal")
+
+    if progress_cached != None and goal_cached != None:
+        progress = progress_cached
+        goal = goal_cached
+    else:
+        challenge_page = http.get(GOODREADS_PROGRESS_URL + CHALLENGE_ID)
+
+        if challenge_page.status_code != 200:
+            fail("Request failed with status %d", challenge_page.status_code)
+
+        body = challenge_page.body()
+        progress_div = re.findall(r"<div class='progressText'>([\s\S]*?)</div>", body)
+
+        if not progress_div:
+            fail("No challenge found at {}".format(config.str("user_challenge_id", CHALLENGE_ID)))
+
+        progress_nums = re.findall(r"\d+", progress_div[0])
+        progress = progress_nums[0]
+        goal = progress_nums[1]
+        cache.set("progress", str(progress), ttl_seconds = 86400)
+        cache.set("goal", str(goal), ttl_seconds = 86400)
+
+    progress_text = " ".join(["Read:", str(progress), "books."])
+    goal_text = " ".join(["Goal:", str(goal), "books."])
+
+    images = gen_book_image(int(progress))
+
+    lines = [
+        render.Text(content = progress_text),
+        render.Text(content = goal_text),
+        render.Padding(child = images, pad = (0, 1, 0, 0)),
+    ]
+
+    return render.Root(
+        child = render.Column(
+            main_align = "space_around",
+            children = lines,
+        ),
+    )
+
+def get_schema():
+    return schema.Schema(
+        version = "1",
+        fields = [
+            schema.Text(
+                id = "user_challenge_id",
+                name = "Challenge ID",
+                desc = "Navigate goodreads.com/user_challenges and get this from the URL",
```

Comment:
This description field is unfortunately not visible to users in the Tidbyt smartphone app, so you probably want to add a note about this to the app description.

Reply:
Great, can do.